### ISSUE TITLE
Question: does this library support dependencies between tasks ?

### ISSUE THREAD

**Issue by boxerab** [dougbinks/enkiTS#29]
Nice library! 
Suppose I have a task `task123` that can only execute when `task1`, `task2` and `task3` have completed.  Can enkiTS support this work flow ?

**Comment by dougbinks** [comment]
enkiTS supports two approaches to dependencies - tasks can spawn other tasks and tasks can wait on tasks to complete. Explicit dependencies are in the works, but I don't have an eta.

In your example one approach might be to use ```WaitforTask``` to wait for each task in turn. ```WaitforTask``` will run any task which can be executed whilst waiting.

```
// This code may work but could create deadlocks
struct task123: enki::ITaskSet {
   virtual void    ExecuteRange(  enki::TaskSetPartition range, uint32_t threadnum ) {
        g_TS.WaitforTask( &task1);
        g_TS.WaitforTask( &task2);
        g_TS.WaitforTask( &task3);
        // do work
   }
};
```
However do note that this above code could lead to deadlock if other tasks are also using wait, either due to A waiting on B and B waiting on A or due to the function WaitforTask running another task which waits on the task which called WaitforTask.

To avoid deadlocks I recommend only waiting on tasks which your task/thread has spawned - this may require creating a task to spawn ```task1```, ```task2```, and ```task3``` then wait before spawning ```task123```:

```
struct TaskSpawn123: enki::ITaskSet {
   virtual void    ExecuteRange(  enki::TaskSetPartition range, uint32_t threadnum ) {
        g_TS.AddTaskSetToPipe( &task1 );
        g_TS.AddTaskSetToPipe( &task1 );
        g_TS.AddTaskSetToPipe( &task1 );
        g_TS.WaitforTask( &task1);
        g_TS.WaitforTask( &task2);
        g_TS.WaitforTask( &task3);
        g_TS.AddTaskSetToPipe( &task123);
   }
};
```

**Comment by boxerab** [comment]
Thanks. so, as long as the task graph has no cycles, I should be ok. 
Btw, I am trying to replicate the functionality for OpenCL [events](http://www.heterogeneouscompute.org/hipeac2011Presentations/OpenCL-events.pdf), although
in that case, there is both a hardware and a software queue.

**Comment by boxerab** [comment]
Would it make sense to decouple the events from the tasks? So that tasks wait for events to signal complete, rather than tasks ? Perhaps you already do this in the library internals.

**Comment by boxerab** [comment]
Also, is it possible for WaitForTask to call a callback when complete, so all tasks could be enqueued at one time, and the scheduler would handle dependencies at run time ?

**Comment by boxerab** [comment]
And also, could WaitForTask wait for multiple tasks?  These are just some ideas I had, not feature requests ! :)

**Comment by dougbinks** [comment]
Re: events, this is similar to the explicit dependencies which are a work in progress.

For WaitForTask to call a callback you can implement that yourself by creating a task which launches the  tasks, waits for them, and then calls the callback or if you need the callback on one particular thread you can use the pinned tasks.

Waiting for multiple tasks is done reasonably efficiently by waiting for them one at a time as I've shown above. If you have hundreds of tasks then you might consider whether you actually have a tasksets, i.e. several data parallel tasks rather than lots of single tasks - see the [ParallelSum example](https://github.com/dougbinks/enkiTS/blob/master/example/ParallelSum.cpp) for a data parallel task set.

**Comment by dougbinks** [comment]
Glad the example code helps! Let me know if you have any other questions.
